### DISCUSSION TITLE
Which features of C++ consume most memory in clangd LSP?

### DISCUSSION THREAD

**Discussion by aerosayan** [clangd/clangd#1928] [General]
Dear Developers, 

Thanks for creating and maintaining clangd.

I wanted to ask, if there's a _**formal list or benchmark**_ of which features in C++ consume most memory in clangd.

Because, currently not everyone of our team has access to powerful computers, or build servers, so during our last project, using heavily templated libraries like Eigen3, AWS S3 SDK etc. consumed too much memory (>4 GB), and caused severe development problems for us.

Mainly templates seem to cause problems, as they're instantiated in every file they're included in, and consume too much memory.

So, for our future work, we're either going to not use templates, or compile the templated code as a separate translation unit or library, and access them through a C like interface using raw pointers, so the LSP doesn't consume so much memory.

We're thinking of using a more sane subset of C++ (possibly C++11, without any templates) that won't take too much time to compile, or consume too much memory in a LSP. 

"Orthodox C++" is a good style guide we're trying to follow, with this regard, which is mostly "C like C++": https://gist.github.com/bkaradzic/2e39896bc7d8c34e042b

But it would be nice to know before hand, which features in C++ require most memory for the LSP.

Thanks
~Sayan

PS:

I benchmarked clangd's performance with C (99) code, and it works very well. 
- For a file of 100-400 lines long C file, clangd consumed max 90 MB.
- But mostly, for smaller files, clangd consumed less than 50 MB.

I also benchmarked clangd's perormance with C++ (11), and if we don't include templates, it seems to work well.
- For a file of 100-400 lines long C++ file that doesn't use templates, clangd consumed max 150 MB.
- But mostly, for smaller files, clangd consumed less than 100 MB.

**Comment by HighCommander4** [comment]
See also https://github.com/clangd/clangd/discussions/1469

**Comment by HighCommander4** [comment]
> I wanted to ask, if there's a _**formal list or benchmark**_ of which features in C++ consume most memory in clangd.

Not that I'm aware of.

It's not immediately clear to me how such a benchmark might be created, but if you have ideas (and especially if you're interested in creating such a benchmark), please feel free to use this issue to discuss it.
